test_runner/regress/test_compatibility.py

```diff
@@ -247,12 +248,29 @@ def test_forward_compatibility(
 
         neon_env_builder.start()
 
+        # not using env.pageserver.version because it was initialized before
+        prev_pageserver_version_str = env.get_binary_version("pageserver")
+        prev_pageserver_version_match = re.search(
+            "Neon page server git-env:(.*) failpoints: (.*), features: (.*)",
+            prev_pageserver_version_str,
+        )
+        if prev_pageserver_version_match is not None:
+            prev_pageserver_version = prev_pageserver_version_match.group(1)
+        else:
+            raise AssertionError(
+                "cannot find git hash in the version string: " + prev_pageserver_version_str
+            )
+
         check_neon_works(
             env,
             test_output_dir=test_output_dir,
             sql_dump_path=compatibility_snapshot_dir / "dump.sql",
             repo_dir=env.repo_dir,
         )
+
+        # ensure the specified pageserver is running
+        assert env.pageserver.log_contains("git-env:" + prev_pageserver_version)
```

Comment:
What ensures that the compatibility snapshot doesn't include the pageserver log from the time it was created, i.e., with the old version?

I think checking for log is quite brittle.

At a minimum, let's assert that before the `neon_env_builder.start()` above,  the log doesn't contain `prev_pageserver_version`

Reply:
Compatibility test runs in two steps: create_snapshot test, and this test. In create_snapshot test, it uses the current version. In forward_compat test, it uses the old version. They produce logs in two different directories, and therefore this should not include the log from the create snapshot step.

Moved with https://github.com/neondatabase/neon/pull/7685/commits/cec385223405b39d9a01b68219655f2c4e72756f

I feel checking the log is the best solution because it reflects "what is actually running". There could be bugs with the test itself, or the neon_local thing, or whatever else, that caused the running pageserver version not as expected.